### PR TITLE
Add `Nep5Token.Burn()`

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Comment by shargon** [comment]
It will be used for neo or gas?

**Comment by erikzhang** [comment]
It will be used for gas.

File: neo/SmartContract/Native/Tokens/Nep5Token.cs [hunk 1]
```diff
@@ -73,6 +73,32 @@ internal protected virtual void MintTokens(ApplicationEngine engine, UInt160 acc
...
+            else
+            {
+                state.Balance -= amount;
+                storage.Value = state.ToByteArray();
+            }
+            storage = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_TotalSupply));
+            BigInteger totalSupply = new BigInteger(storage.Value);
+            totalSupply -= amount;
+            storage.Value = totalSupply.ToByteArray();
+            engine.SendNotification(ScriptHash, new StackItem[] { "Transfer", account.ToArray(), StackItem.Null, amount });
```
Comment (shargon) [inline neo/SmartContract/Native/Tokens/Nep5Token.cs]: Maybe we can use specific notification for mint and burn, no reuse the Transfer

Comment (erikzhang) [reply neo/SmartContract/Native/Tokens/Nep5Token.cs]: https://github.com/neo-project/proposals/blob/master/nep-5.mediawiki#transfer-1

> A token contract which creates new tokens MUST trigger a transfer event with the from address set to null when tokens are created.

>A token contract which burns tokens MUST trigger a transfer event with the to address set to null when tokens are burned.

**Comment by erikzhang** [comment]
> How to call the function? Should add this in the main?

It will be used when process the fees.

**Comment by vncoelho** [comment]
O.o, burned 4rever?
Ripple uses similar strategy.

**Comment by erikzhang** [comment]
The mechanism is not changed from NEO 2.x.

File: neo/SmartContract/Native/Tokens/Nep5Token.cs [hunk 2]
```diff
@@ -73,6 +73,32 @@ internal protected virtual void MintTokens(ApplicationEngine engine, UInt160 acc
             engine.SendNotification(ScriptHash, new StackItem[] { "Transfer", StackItem.Null, account.ToArray(), amount });
         }
 
+        internal protected virtual void Burn(ApplicationEngine engine, UInt160 account, BigInteger amount)
+        {
+            if (amount.Sign < 0) throw new ArgumentOutOfRangeException(nameof(amount));
+            if (amount.IsZero) return;
+            StorageKey key = CreateAccountKey(account);
+            StorageItem storage = engine.Snapshot.Storages.GetAndChange(key);
```
Comment (shargon) [inline neo/SmartContract/Native/Tokens/Nep5Token.cs]: without factory, this will throw an exception, i think that this is expected

Comment (erikzhang) [reply neo/SmartContract/Native/Tokens/Nep5Token.cs]: Yes, this is expected.

**Pull request merged**